packages/core/src/dispatcher.ts

```diff
@@ -58,8 +62,70 @@ export default class Dispatcher extends EventEmitter {
    */
   async init(): Promise<void> {
     this._peerId = this._peerId || (await this._ipfs.id()).id
-    await this._ipfs.pubsub.subscribe(this.topic, this.handleMessage.bind(this))
-    this._log({ peer: this._peerId, event: 'subscribed', topic: this.topic })
+    await this._subscribe()
+    !TESTING && this._resubscribeOnDisconnect()
+  }
+
+  /**
+   * Subscribes IPFS pubsub to `this.topic` and logs an `subscribe` event.
+   */
+  async _subscribe(): Promise<void> {
+    if (this._isSubscribed) {
+      console.log(`Pubsub subscribe skipped. Already subscribed to topic ${this.topic}`)
+    } else {
+      try {
+        await this._ipfs.pubsub.subscribe(
+          this.topic,
+          this.handleMessage.bind(this),
+          {timeout: !TESTING && IPFS_GET_TIMEOUT}
+        )
+
+        const { isSubscribed, error } = await this._confirmIsSubscribed()
+        if (isSubscribed) {
+          this._log({peer: this._peerId, event: 'subscribed', topic: this.topic })
+        } else if (error) {
+          throw error
+        } else {
+          throw new Error(`Pubsub subscribe failed for topic ${this.topic}`)
+        }
+      } catch (error) {
+        console.error(error)
+      }
+    }
+  }
+
+  /**
+   * Periodically checks that IPFS pubsub is subscribed to `this.topic` and
+   * attempts to subscribe if not.
+   */
+  _resubscribeOnDisconnect(): void {
+    this._resubscribeInterval = setInterval(async () => {
+      const { isSubscribed } = await this._confirmIsSubscribed()
+      !isSubscribed && await this._subscribe()
+    }, IPFS_RESUBSCRIBE_INTERVAL_DELAY)
+  }
+
+  /**
+   * Sets `this._iSubscribed` if the `this.topic` is listed as an IPFS pubsub
+   * subscription.
+   * @returns A tuple of `this._isSubscribed` and an error if one was caught.
+   */
+  async _confirmIsSubscribed(): Promise<{isSubscribed: boolean, error: Error | null}> {
+    let isSubscribed = true
+    let error = null
+
+    try {
+      const subscriptions = await this._ipfs.pubsub.ls({ timeout: IPFS_GET_TIMEOUT })
```

Comment:
is there any harm to attempting to subscribe when you're already subscribed?  If not, then we could just always try to subscribe, and we could remove this entire method and keep the logic pretty simple

Reply:
I think it might throw an error saying that the particular listener is already added, so could be tested. But the way it's done now seems fine imo, it will only be one more call if we're not subscribed.